daemon/cluster/cluster.go

```diff
@@ -520,55 +380,50 @@ func (c *Cluster) Join(req types.JoinRequest) error {
 		}
 	}
 
-	// todo: check current state existing
-	n, err := c.startNewNode(nodeStartConfig{
+	clearPersistentState(c.root)
+
+	nr, err := c.newNodeRunner(nodeStartConfig{
 		RemoteAddr:    req.RemoteAddrs[0],
 		ListenAddr:    net.JoinHostPort(listenHost, listenPort),
 		AdvertiseAddr: advertiseAddr,
 		joinAddr:      req.RemoteAddrs[0],
 		joinToken:     req.JoinToken,
 	})
 	if err != nil {
-		c.Unlock()
 		return err
 	}
-	c.Unlock()
+
+	c.mu.Lock()
+	c.nr = nr
+	c.mu.Unlock()
 
 	select {
 	case <-time.After(swarmConnectTimeout):
-		// attempt to connect will continue in background, but reconnect only if it didn't fail
-		go func() {
-			select {
-			case <-n.Ready():
-				c.reconnectOnFailure(n)
-			case <-n.done:
-				logrus.Errorf("failed to join the cluster: %+v", c.err)
-			}
-		}()
 		return ErrSwarmJoinTimeoutReached
-	case <-n.Ready():
-		go c.reconnectOnFailure(n)
-		return nil
-	case <-n.done:
-		c.RLock()
-		defer c.RUnlock()
-		return c.err
+	case err := <-nr.Ready():
+		if err != nil {
+			c.mu.Lock()
+			c.nr = nil
+			c.mu.Unlock()
+		}
+		return err
 	}
 }
 
 // GetUnlockKey returns the unlock key for the swarm.
 func (c *Cluster) GetUnlockKey() (string, error) {
-	c.RLock()
-	defer c.RUnlock()
+	c.mu.RLock()
+	defer c.mu.RUnlock()
 
-	if !c.isActiveManager() {
-		return "", c.errNoManager()
+	state := c.currentNodeState()
+	if !state.IsActiveManager() {
```

Comment:
We are working on a snapshot of a last known state in here. `handleControlSocketChange` would switch the runner internals so next call to `currentNodeState` gets the correct value. Same for next comment.